### PR TITLE
WIP: Loom support

### PR DESCRIPTION
**Pull request by jdegoes**
Add experimental support for Loom, including:

1. Using the default Loom scheduler
2. Backing every fiber with a virtual thread (guaranteed 1-to-1 mapping)
3. Embracing blocking wherever relevant, e.g. when converting a Java `Future` to a ZIO effect
4. Eliminating or minimizing `synchronized` in favor of non-OS-thread blocking constructs like `Lock`
5. Introducing dual code paths for every `ZIO.async`, such that `async` is no longer used under Loom
6. Refining ZIO semantics in cases where that's necessary (e.g. green threads cannot shift!)

### CODE REVIEW

File: core/jvm/src/main/scala/zio/interop/javaz.scala [hunk 1]
```diff
@@ -95,11 +95,25 @@ private[zio] object javaz {
...
+          try thunk.get()
+          catch {
+            case e: ExecutionException => throw e.getCause
+          }
+        } else {
+          var result: A = null.asInstanceOf[A]
+
+          while (result == null) {
+            try {
+              result = thunk.get()
```
Comment (guersam) [inline core/jvm/src/main/scala/zio/interop/javaz.scala:110]: Wouldn't it make the effect run multiple times due to the re-evaluation of the side effecting named parameter `=> Future[A]`?

**Comment by WtzLAS** [comment]
any further progress on this?

**Comment by domdorn** [comment]
as just mentioned in the spartan session, maybe this
https://index.scala-lang.org/sbt/sbt-multi-release-jar
could allow you to ship one jar with different behavior for loom-aware and non loom-aware JVMs. 
E.g. I'm just upgrading all our services to JDK21, so we could already use this there.

**Comment by devcdcc** [comment]
Hello @jdegoes @adamgfraser  any update about it? 
:c

**Comment by jdegoes** [comment]
Closing this, as most of this work was already merged in preparation for 2.1.

**Pull request closed without merging**